### PR TITLE
Cosmos: Add API to configure partition keys and use them in the update pipeline.

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Fix Northwind tests

Part of #12086

### CODE REVIEW

File: test/EFCore.Cosmos.FunctionalTests/TestUtilities/CosmosTestStore.cs [hunk 1]
```diff
@@ -115,9 +116,8 @@ private async Task CreateFromFile(DbContext context)
 
                                                             document["id"] = $"{entityName}|{document["id"]}";
                                                             document["Discriminator"] = entityName;
-                                                            document["__partitionKey"] = "0";
 
-                                                            await cosmosClient.CreateItemAsync("NorthwindContext", document);
+                                                            await cosmosClient.CreateItemAsync("NorthwindContext", document, null);
```
Comment (smitpatel) [inline test/EFCore.Cosmos.FunctionalTests/TestUtilities/CosmosTestStore.cs]: Is it still NorthwindContext rather than entityType name?

Comment (AndriySvyryd) [reply test/EFCore.Cosmos.FunctionalTests/TestUtilities/CosmosTestStore.cs]: Yes, there was a bug in model building that didn't configure this correctly.

File: src/EFCore.Cosmos/Extensions/CosmosEntityTypeBuilderExtensions.cs [hunk 2]
```diff
@@ -160,5 +163,96 @@ public static class CosmosEntityTypeBuilderExtensions
...
+        ///     Configures the property that is used to store the partition key.
+        /// </summary>
+        /// <param name="entityTypeBuilder"> The builder for the entity type being configured. </param>
+        /// <param name="name"> The name of the partition key property. </param>
+        /// <returns> The same builder instance so that multiple calls can be chained. </returns>
+        public static EntityTypeBuilder ForCosmosHasPartitionKey(
+            [NotNull] this EntityTypeBuilder entityTypeBuilder,
+            [CanBeNull] string name)
+        {
+            entityTypeBuilder.Metadata.SetCosmosContainingPropertyName(name);
```
Comment (smitpatel) [inline src/EFCore.Cosmos/Extensions/CosmosEntityTypeBuilderExtensions.cs]: Wrong method

File: src/EFCore.Cosmos/Extensions/CosmosEntityTypeBuilderExtensions.cs [hunk 3]
```diff
@@ -160,5 +163,96 @@ public static class CosmosEntityTypeBuilderExtensions
...
+        /// <param name="propertyExpression"> The  partition key property. </param>
+        /// <returns> The same builder instance so that multiple calls can be chained. </returns>
+        public static EntityTypeBuilder<TEntity> ForCosmosHasPartitionKey<TEntity, TProperty>(
+            [NotNull] this EntityTypeBuilder<TEntity> entityTypeBuilder,
+            [NotNull] Expression<Func<TEntity, TProperty>> propertyExpression)
+            where TEntity : class
+        {
+            Check.NotNull(propertyExpression, nameof(propertyExpression));
+
+            entityTypeBuilder.Metadata.SetCosmosPartitionKeyPropertyName(propertyExpression.GetPropertyAccess().Name);
```
Comment (smitpatel) [inline src/EFCore.Cosmos/Extensions/CosmosEntityTypeBuilderExtensions.cs]: GetSimplerMemberName

File: src/EFCore.Cosmos/Extensions/CosmosEntityTypeBuilderExtensions.cs (hunk 2 above)
Comment (smitpatel) [reply src/EFCore.Cosmos/Extensions/CosmosEntityTypeBuilderExtensions.cs]: Perhaps need a regression test for this

---
In reply to: [295445661](https://github.com/aspnet/EntityFrameworkCore/pull/16148#discussion_r295445661) [](ancestors = 295445661)

File: src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs [hunk 4]
```diff
@@ -184,38 +190,43 @@ public bool DeleteDatabase()
...
 
         private async Task<bool> ReplaceItemOnceAsync(
             DbContext _,
-            (string ContainerId, string ItemId, JObject Document) parameters,
+            (string ContainerId, string ItemId, JObject Document, object PartitionKey) parameters,
             CancellationToken cancellationToken = default)
         {
-            using (var stream = new MemoryStream())
-            using (var writer = new StreamWriter(stream, new UTF8Encoding(), bufferSize: 1024, leaveOpen: false))
+            using var stream = new MemoryStream();
```
Comment (smitpatel) [inline src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs]: :trollface:

File: src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs [hunk 5]
```diff
@@ -184,38 +190,43 @@ public bool DeleteDatabase()
...
+            using var writer = new StreamWriter(stream, new UTF8Encoding(), bufferSize: 1024, leaveOpen: false);
             using (var jsonWriter = new JsonTextWriter(writer))
             {
                 JsonSerializer.Create().Serialize(jsonWriter, parameters.Document);
                 await jsonWriter.FlushAsync(cancellationToken);
 
                 var container = Client.GetDatabase(_databaseId).GetContainer(parameters.ContainerId);
+                var partitionKey = parameters.PartitionKey == null
```
Comment (smitpatel) [inline src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs]: DRY, make private function

**Comment by smitpatel** [comment]
Can you rebase on latest master and make sure SimpleQueryTests are passing?

**Comment by AndriySvyryd** [comment]
@smitpatel  `Where_subquery_anon` is still failing. Is that expected?

**Comment by smitpatel** [comment]
Disable the test in Cosmos

**Comment by smitpatel** [comment]
Those are the tests @maumar enabled in core so need to disable them in cosmos.

**Pull request merged**